src/Build/BackEnd/Components/RequestBuilder/Lookup.cs

```diff
@@ -668,6 +669,7 @@ internal void AddNewItemsOfItemType(string itemType, ICollection<ProjectItemInst
                     itemsToAdd = itemsToAdd.Where(item => !existingItems.Contains(item, ProjectItemInstance.EqualityComparer));
```

Comment:
ooof, now that I'm reading this this is returning an enumerable that will evaluate every time you enumerate it. This is not good. I think this whole method needs to be carefully reviewed (and we'll likely have to allocate to eagerly realize the collection once, because keeping it lazy will have multiple consumers each iterating on it)

Reply:
wait, how does that Contains method work? Is that O(n)?

The more I look the more perf questions I have. The whole method looks super sketchy now. Distinct is expensive, GetItems is expensive, we don't use a Hashtable (??) and we potentially double-enumerate the itemsToAdd

I think the whole method needs to be carefully revisited and we probably need to measure allocations before and after

Reply:
Do we care about the allocations only due to performance constraints or is there a memory reason as well please?
Because if we only care about the performance, then removing the O(n^2) easily justifies an extra hash table allocation. If this would still be an issue, I would probably have to rewrite GetItems() so that it can return a HashSet instead of the current Linked list if required. (Or we would have to really carefully check if we can just return a HashSet every time although that sounds risky)

Reply:
Understood, thanks for the clarification.
Then the question is:
Should I rewrite significant portion of the `GetItems` method to directly return a HashSet to save the allocation in this specific case? (I'm not sure if my understanding is correct, but since ProjectItemInstance is a reference-based object, it should be a hash set with an overhead number of items * pointer size or something which is something but maybe not enough considering the size of the ProjectItemInstances we operate with)

I can do it and I'm fairly sure I now have the context for the change. However it will be fairly decent chunk of code that is affected which is also something to consider.